File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int calculateTotalDays(vector<int>& weights, int capacity) {
        int days = 1;
        int load = 0;
        for (int i = 0; i < weights.size(); i++) {
            if (load + weights[i] > capacity) {
                days = days + 1;
                load = weights[i];
            } else
                load += weights[i];
        }
        return days;
    }

    int shipWithinDays(vector<int>& weights, int days) {
        int low = INT_MIN, high = 0;
        for (auto i : weights) {
            low = max(low, i);
            high += i;
        }
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (calculateTotalDays(weights, mid) <= days) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.java
class Solution {
  Boolean calculate(int[] weights, int days, int weight) {
        int checkDays = 1;
        int sum = 0;
        for (int i = 0; i < weights.length; i++) {

            if (sum + weights[i] <= weight) {
                sum += weights[i];
            } else {
                checkDays++;
                sum = 0;
                sum += weights[i];
            }
        }
        return checkDays <= days;
    }

    public int shipWithinDays(int[] weights, int days) {
        int low = Arrays.stream(weights).max().getAsInt();
        int high = Arrays.stream(weights).sum();
        int ans = high;
        while (low <= high) {
            int mid = (low + high) >> 1;
            if (calculate(weights, days, mid)) {
                ans = mid;
                high = mid - 1;
            } else
                low = mid + 1;
        }
        return ans;
    }
}